library/x509_crt.c

```diff
@@ -620,231 +850,584 @@ static int x509_get_ext_key_usage( unsigned char **p,
+                                         unsigned char *data,
+                                         size_t data_len )
 {
-    int ret;
-    size_t len, tag_len;
-    mbedtls_asn1_buf *buf;
-    unsigned char tag;
-    mbedtls_asn1_sequence *cur = subject_alt_name;
-
-    /* Get main sequence tag */
-    if( ( ret = mbedtls_asn1_get_tag( p, end, &len,
-            MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE ) ) != 0 )
-        return( MBEDTLS_ERR_X509_INVALID_EXTENSIONS + ret );
-
-    if( *p + len != end )
-        return( MBEDTLS_ERR_X509_INVALID_EXTENSIONS +
-                MBEDTLS_ERR_ASN1_LENGTH_MISMATCH );
+    mbedtls_asn1_sequence **cur_ptr = (mbedtls_asn1_sequence **) ctx;
+    mbedtls_asn1_sequence *cur = *cur_ptr;
 
-    while( *p < end )
+    /* Allocate and assign next pointer */
+    if( cur->buf.p != NULL )
     {
-        if( ( end - *p ) < 1 )
-            return( MBEDTLS_ERR_X509_INVALID_EXTENSIONS +
-                    MBEDTLS_ERR_ASN1_OUT_OF_DATA );
-
-        tag = **p;
-        (*p)++;
-        if( ( ret = mbedtls_asn1_get_len( p, end, &tag_len ) ) != 0 )
-            return( MBEDTLS_ERR_X509_INVALID_EXTENSIONS + ret );
-
-        if( ( tag & MBEDTLS_ASN1_TAG_CLASS_MASK ) !=
-                MBEDTLS_ASN1_CONTEXT_SPECIFIC )
-        {
-            return( MBEDTLS_ERR_X509_INVALID_EXTENSIONS +
-                    MBEDTLS_ERR_ASN1_UNEXPECTED_TAG );
-        }
-
-        /* Skip everything but DNS name */
-        if( tag != ( MBEDTLS_ASN1_CONTEXT_SPECIFIC | 2 ) )
-        {
-            *p += tag_len;
-            continue;
-        }
-
-        /* Allocate and assign next pointer */
-        if( cur->buf.p != NULL )
-        {
-            if( cur->next != NULL )
-                return( MBEDTLS_ERR_X509_INVALID_EXTENSIONS );
-
-            cur->next = mbedtls_calloc( 1, sizeof( mbedtls_asn1_sequence ) );
-
-            if( cur->next == NULL )
-                return( MBEDTLS_ERR_X509_INVALID_EXTENSIONS +
-                        MBEDTLS_ERR_ASN1_ALLOC_FAILED );
-
-            cur = cur->next;
-        }
-
-        buf = &(cur->buf);
-        buf->tag = tag;
-        buf->p = *p;
-        buf->len = tag_len;
-        *p += buf->len;
+        cur->next = mbedtls_calloc( 1, sizeof( mbedtls_asn1_sequence ) );
+        if( cur->next == NULL )
+            return( MBEDTLS_ERR_ASN1_ALLOC_FAILED );
+        cur = cur->next;
     }
 
-    /* Set final sequence entry's next pointer to NULL */
-    cur->next = NULL;
-
-    if( *p != end )
-        return( MBEDTLS_ERR_X509_INVALID_EXTENSIONS +
-                MBEDTLS_ERR_ASN1_LENGTH_MISMATCH );
+    cur->buf.tag = tag;
+    cur->buf.p = data;
+    cur->buf.len = data_len;
 
+    *cur_ptr = cur;
     return( 0 );
 }
 
+static int x509_get_subject_alt_name( unsigned char *p,
+                                      const unsigned char *end,
+                                      mbedtls_x509_sequence *subject_alt_name )
+{
+    return( mbedtls_asn1_traverse_sequence_of( &p, end,
```

Comment:
I am aware and agree that we'll support more name types in the future, but the current form deliberatlely maintains the behaviour as we have it now.

As discussed in #2530, I think it would make sense to drop the restriction altogether, and to return a list of all SAN components (unparsed - as justified above), but I'm not sure that should be addressed here.

Reply:
No change.

Reply:
>I am aware and agree that we'll support more name types in the future, but the current form deliberatlely maintains the behaviour as we have it now.

I disagree. In user point of view, yes this is same behavior. However, in current behavior, if you want to enhance the type parsing support, you need to add the support in `x509_get_subject_alt_name`. Now, you will need to change the callback altogether, and replace the call to `mbedtls_asn1_traverse_sequence_of` to something else

Reply:
Please see my extended reply above on how I propose to extend the parsing: We must maintain an `mbedtls_x509_sequence` of raw SAN components (whether this restricts to dNSName or includes some/all other names doesn't matter for this) for API stability / ease of API progression, and further parsing should come on top.

Reply:
Since we are introducing new API, I don't see a use case of API stability and ease of pregression.
AS mentioned, I don't mind us keeping the internal functionality as is, but the high level public API should change IMHO